### DISCUSSION TITLE
LSP in neovim

### DISCUSSION THREAD

**Discussion by Amzd** [AparokshaUI/Adwaita#31] [Q&A] [answered]
I would prefer to use nvim over Builder. 

With the template app I used `flatpak run org.flatpak.Builder build <app id>.json` and this works but the swift build artifacts aren't saved to .build folder which means no incremental builds and no LSP.

I have been trying to just manually copy them to the .build folder but I am not able to get incremental builds working this way. I have put in a couple hours trying to get this working so I thought I should probably reach out instead. 

Is there a command for flatpak-builder that just uses the current directory as build dir or is GNOME Builder using something different? Any info is greatly appreciated as I am new to flatpak.

**Comment by Amzd** [comment, marked as answer]
Okay after some more experimentation and research I landed on this manifest

```json
{
  ...
  "build-options": {
    ...
    "build-args": [
      "--share=network" 
      "--build-dir=<<PROJ_DIR>>",
      "--filesystem=<<PROJ_DIR>>",
    ],
    "env": {
      "FLATPAK_BUILDER_BUILDDIR": "<<PROJ_DIR>>"
    }
  },
```
- `--share=network` because SPM needs to fetch dependencies (you can omit this if you run `swift package resolve` manually)
- `--filesystem=` gives the build system access to project files instead of only /run/build/proj
- `--build-dir=` makes sure the `swift build` command is run inside the project folder instead of the sandboxed /run/build/proj
- `FLATPAK_BUILDER_BUILDDIR` by default `flatpack-builder` sets this to whatever folder it tries to build in so if you rely on this you should update it to your proj folder.

This allows for incremental builds but **not** LSP support.

I currently have this set in a `debug.com.example.Project.json` manifest file so I don't add the project dirs to the actual final release.

Makefile
```mk
debug:
	@flatpak-builder .build/flatpak-builder debug.me.amzd.Recommended.json --force-clean --install --user
	@flatpak run me.amzd.Recommended

release:
	@flatpak-builder .build/flatpak-builder me.amzd.Recommended.json --force-clean --install --user

install_dependencies:
	@flatpak install org.freedesktop.Sdk.Extension.swift5 
```

**Reply by Amzd** [reply]
To get LSP working I locally installed `libadwaita-devel` which feels like it is defeating the purpose of flatpak a bit but it works.

**Reply by david-swift** [reply]
With GNOME Builder, all of that works fine without editing the manifest or installing Libadwaita, so it might make sense to figure out what commands GNOME Builder uses. Thanks for the detailed instructions for others interested in using nvim until we (hopefully) find a more elegant way.

**Reply by Amzd** [reply]
> it might make sense to figure out what commands GNOME Builder uses

Yea I couldn't figure out how to inspect these in GNOME Builder but I had a look at the VSC*de plugin implementation and the build logic was quite a bit more complicated than I wanted to make myself, so `flatpak-builder` was really the only option. I might look at how they run the LSP if I have more trouble in the future with dependencies.
